poseidon2-circuit-air/src/air.rs

```diff
@@ -201,40 +206,71 @@ impl<
 
             // Apply chaining rules.
             // NOTE: For rows with new_start = false:
-            // - Sponge mode (merkle_path = 0): all limbs come from the previous output unless
-            //   a limb is exposed via in_ctl, in which case the provided input overrides it.
-            // - Merkle mode (merkle_path = 1): the previous digest is always the previous output limbs 0-1.
-            //   If mmcs_bit = 0 (previous digest is left child), chain into input limbs 0-1; input limbs 2-3 come from inputs.
-            //   If mmcs_bit = 1 (previous digest is right child), chain into input limbs 2-3; input limbs 0-1 come from inputs.
+            // - In sponge mode (merkle_path = 0), the Poseidon input state is *entirely*
+            //   taken from the previous row's output; `input_values` are unused for chained limbs.
+            // - In Merkle mode (merkle_path = 1), only rate limbs (0..RATE_EXT-1) are chained;
+            //   capacity limbs come from `input_values`. This matches the Poseidon Perm Table spec.
             // - If in_ctl[i] = 1, that limb is NOT chained and comes from CTL/witness instead.
             //   The AIR constraints will enforce this (chaining is gated by 1 - in_ctl[i]).
             let mut state = padded_inputs;
             let i = inputs.len();
             if i > 0 && !*new_start {
                 if *merkle_path {
-                    // Merkle-path mode: the previous digest is always the previous row's out[0..1].
-                    // `mmcs_bit` selects whether that digest is the left (0) or right (1) child:
-                    // - bit=0: chain into input limbs 0..1
-                    // - bit=1: chain into input limbs 2..3
+                    // Merkle-path mode: chain based on this row's mmcs_bit
+                    // Only chain rate limbs (0..RATE_EXT-1) if in_ctl[i] = 0 (handled by AIR constraints)
                     if let Some(prev_out) = prev_output {
-                        if !*mmcs_bit {
-                            if !in_ctl[0] {
-                                state[0..D].copy_from_slice(&prev_out[0..D]);
-                            }
-                            if !in_ctl[1] {
-                                state[D..2 * D].copy_from_slice(&prev_out[D..2 * D]);
+                        let cur_bit = *mmcs_bit;
+                        // For trace generation, we chain unconditionally here.
+                        // The AIR will enforce that chaining only applies when in_ctl = 0.
+                        if cur_bit {
+                            // Case B: mmcs_bit = 1 (right = previous hash)
+                            // If RATE_EXT >= 2*DIGEST_EXT, chain from second half of rate section
+                            // Otherwise, fall back to capacity limbs
+                            if RATE_EXT >= 2 * DIGEST_EXT {
+                                // Chain from second half of rate: in_{r+1}[0..DIGEST_EXT-1] = out_r[DIGEST_EXT..2*DIGEST_EXT-1]
```

Comment:
I think the source is still `out_r[0..DIGEST_EXT]`, but it's copied into `in_{r+1}[DIGEST_EXT..2*DIGEST_EXT-1]`. I'm surprised it's not caught by any test.

I'm not sure how pertinent it is to correct it though, since the state computation logic will be removed from here in https://github.com/Plonky3/Plonky3-recursion/pull/222.